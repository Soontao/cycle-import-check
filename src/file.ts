import { concat, filter, isArray, isString, join as arrayJoin, keys, map } from "@newdash/newdash";
import { includes } from "@newdash/newdash/includes";
import { readFileSync } from "fs";
import { sync } from "glob";
import { dirname, join, join as pathJoin, normalize, relative } from "path";
import { cwd } from "process";
import { Extension, FileImportDescription, PackageJson, ReportVO, ScanResult } from "./type";

require.extensions[".ts"] = require.extensions[".js"]
require.extensions[".jsx"] = require.extensions[".js"]
require.extensions[".tsx"] = require.extensions[".js"]
require.extensions[".mjs"] = require.extensions[".js"]

const { resolve } = require

export const allDependencies = (absPath: string) => {
  return concatAllDependencies(findProjectPackageJson(absPath))
}

export const concatAllDependencies = (json: PackageJson): string[] => {
  try {
    const { dependencies, devDependencies, peerDependencies } = json;
    var rt = [];
    if (dependencies) {
      rt = concat(rt, keys(dependencies))
    }
    if (devDependencies) {
      rt = concat(rt, keys(devDependencies))
    }
    if (peerDependencies) {
      rt = concat(rt, keys(peerDependencies))
    }
    return rt;
  } catch (error) {
    throw new Error("please run cycle-import-check in npm project (with project.json)")
  }

}

export const findProjectPackageJson = (absPath: string): PackageJson => {
  const finder = require("find-package-json")(absPath)
  return finder.next().value;
}

export const filterNodeDependenciesImport = (descriptions: FileImportDescription[], dependencies: string[]) => {
  // @ts-ignore
  return filter(descriptions, i => !includes(dependencies, i.importFile))
}

/**
 * list all acceptable files in a specific directory
 *
 * @param dir
 * @param ext
 */
export const listAllFile = (dir: string, ext: Extension[] = []) => {
  return sync(pathJoin(dir, `./**/*.{${arrayJoin(ext, ",")}}`), {
    realpath: true,
    ignore: [
      "**/node_modules/**"
    ]
  })
}

/**
 * read file content
 *
 * @param absolutePath
 */
export const readFile = (absolutePath: string) => {
  return readFileSync(absolutePath, { encoding: "utf8" })
}

/**
 * will throw error if file not exist
 *
 * @param fromFileAbsolutePath
 * @param importFileRelativePath
 */
export const resolveFilePath = (fromFileAbsolutePath: string, importFileRelativePath: string) => {
  const dir = dirname(fromFileAbsolutePath);
  const targetPath = join(dir, importFileRelativePath);
  // to do replace nodejs resolve function
  try {
    return normalize(resolve(targetPath));
  } catch (error) {
    // can not resolve import file
    return ""
  }
}

/**
 * map absolute path to relative path
 *
 * @param paths
 */
export const mapAbsPathsToRelPaths = (paths: string | string[]): string | string[] => {
  if (isString(paths)) {
    return relative(cwd(), paths)
  }
  if (isArray(paths)) {
    return map(paths, p => relative(cwd(), p))
  }
}

export const mapScanResultToReportVO = (result: ScanResult): ReportVO => {
  var rt: ReportVO = { nodes: [], links: [] }
  rt.nodes = map(result.nodes, n => ({ name: (mapAbsPathsToRelPaths(n) as string) }))
  rt.links = map(result.imports, i => ({
    source: (mapAbsPathsToRelPaths(i.fromFile) as string),
    target: (mapAbsPathsToRelPaths(i.importFile) as string),
    value: i.code,
  }))
  return rt;
}

