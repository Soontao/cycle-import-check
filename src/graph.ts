import { Graph, alg } from "graphlib";
import { forEach } from "@newdash/newdash";
import { FileImportDescription } from "./type";

/**
 * calculate cycle import file
 *
 * @param fileAbsPathList files absolutely path
 * @param imports each file's import
 */
export const calculateCycleImport = (fileAbsPathList: string[], imports: FileImportDescription[]) => {
  const new_graph = new Graph({ directed: true })
  new_graph.setNodes(fileAbsPathList)
  forEach(imports, i => new_graph.setEdge(i.fromFile, i.importFile))
  return alg.findCycles(new_graph)
}

