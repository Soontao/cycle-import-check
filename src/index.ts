

export { scanDirectoryWithResult } from "./scanner"

export { calculateCycleImport } from "./graph"

