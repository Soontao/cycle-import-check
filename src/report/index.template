<!DOCTYPE html>
<html lang="zh-CN">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>Cycle Dependency Check</title>
  <script src="https://cdn.bootcss.com/echarts/4.1.0/echarts.min.js"></script>
</head>

<body style="width: 100%; height: 100%; position: absolute;">
  <div id="container" style="width: 100%;height: 100%;"></div>
  <script>
    var chart = echarts.init(document.getElementById('container'));

    var data = {{data}}

    var links = {{links}}

    // 指定图表的配置项和数据
    var option = {

      tooltip: {},
      animationDurationUpdate: 200,
      animationEasingUpdate: 'quinticInOut',

      series: [
        {
          type: 'graph',
          layout: 'none',
          symbolSize: 20,
          roam: true,
          layout: "circular",
          focusNodeAdjacency: true,
          label: {
            normal: {
              show: false,
              fontSize: 10,
              position: "right",
            }
          },
          edgeSymbol: ['arrow'],
          edgeSymbolSize: [10],
          edgeLabel: {
            normal: {
              textStyle: {
                fontSize: 20
              }
            }
          },
          tooltip: {
            show: false
          },
          data: data,
          edgeLabel: {
            formatter: function (params) {
              return params.value || "";
            },
            fontSize: 10,
            show: true,
          },
          links: links,
          lineStyle: {
            opacity: 0.9,
            width: 2,
            curveness: 0.1
          }
        }
      ]
    };

    chart.setOption(option);
  </script>
</body>

</html>