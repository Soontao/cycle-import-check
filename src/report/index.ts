import { ReportVO } from "../type";
import { readFileSync } from "fs";
import { join } from "path";

export const generateReportHTML = (report: ReportVO) => {
  const tmpl: string = readFileSync(join(__dirname, "./index.template"), { encoding: "utf8" })
  const withData = tmpl.replace("{{data}}", JSON.stringify(report.nodes, null, 2))
  const withDataAndLink = withData.replace("{{links}}", JSON.stringify(report.links, null, 2))
  return withDataAndLink;
}