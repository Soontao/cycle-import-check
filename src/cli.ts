#!/usr/bin/env node

import "colors"
import { cwd, exit, argv } from "process";
import { error, log } from "console";
import { forEach } from "lodash";
import { scanDirectoryWithResult } from "./scanner";
import { join, isAbsolute } from "path";
import { mapAbsPathesToRelPathes } from "./file";

const wordspaceDir = cwd();

var directory = argv[2]

if (!directory) {
  directory = "."
}

if (!isAbsolute(directory)) {
  directory = join(wordspaceDir, directory)
}

const result = scanDirectoryWithResult(directory)

if (result.haveCycle) {
  error(`Circular dependency existed in ${directory}`.red)
  forEach(result.cyclies, (cycle, index) => {
    error(`\ncycle ${index + 1}, size (${cycle.length}):\n`)
    forEach(mapAbsPathesToRelPathes(cycle), c => error(`  ${c}`.red))
  })
  exit(1)
} else {
  log(`Congratulation, no import cycle founded in ${directory}`.green)
  exit(0)
}