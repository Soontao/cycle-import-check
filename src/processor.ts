import { resolveFilePath } from "./file";
import { FileImportDescription } from "./type";
import {parse} from "@babel/parser";
import traverse from "@babel/traverse";

/**
 * findFileDependencies
 *
 * @param fileAbsolutePath
 * @param fileCodeString
 */
export const findFileDependencies = (fileAbsolutePath: string, fileCodeString: string) => {
  const result: FileImportDescription[] = []
  try {
    const ast = parse(fileCodeString, {
      sourceType: "unambiguous",
      plugins: [
        "typescript",
        "jsx"
      ]
    })
    traverse(ast, {
      ImportDeclaration: (p) => {
        const {node} = p;
        if (node.source) {
          const sourceFile = resolveFilePath(fileAbsolutePath, node.source.value)
          if (sourceFile) {
            result.push({
              fromFile: fileAbsolutePath,
              importFile: sourceFile,
              code: fileCodeString.slice(node.start, node.end)
            })
          }
        }
      },
      ExportNamedDeclaration: (p) => {
        const {node} = p;
        if (node.source) {
          const sourceFile = resolveFilePath(fileAbsolutePath, node.source.value)
          if (sourceFile) {
            result.push({
              fromFile: fileAbsolutePath,
              importFile: sourceFile,
              code: fileCodeString.slice(node.start, node.end)
            })
          }
        }
      },
      ExportAllDeclaration: (p) => {
        const {node} = p;
        if (node.source) {
          const sourceFile = resolveFilePath(fileAbsolutePath, node.source.value)
          if (sourceFile) {
            result.push({
              fromFile: fileAbsolutePath,
              importFile: sourceFile,
              code: fileCodeString.slice(node.start, node.end)
            })
          }
        }
      },
      CallExpression: (p) => {
        const {node} = p;
        if (node.callee.type === "Identifier" && node.callee.name === "require") {
          if (node.arguments.length >= 1 && node.arguments[0].type === "StringLiteral") {
            const sourceFile = resolveFilePath(fileAbsolutePath, node.arguments[0].value)
            if (sourceFile) {
              result.push({
                fromFile: fileAbsolutePath,
                importFile: sourceFile,
                code: fileCodeString.slice(node.start, node.end)
              })
            }
          }
        }
      }
    });
  } catch (error) {
    console.log("Error while parsing", fileAbsolutePath);
    throw error;
  }

  return result;
}
