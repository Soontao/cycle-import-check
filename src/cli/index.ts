#!/usr/bin/env node

import { parse } from "cli";
import { cwd } from "process";
import { error, log } from "console";

const wordspaceDir = cwd();

const options = parse({
  "directory": ["d", "scaned directory", "string"],
  "jsx": ["x", "scan jsx & tsx files", "boolean", false],
  "ts": ["t", "scan ts files", "boolean", false]
})

