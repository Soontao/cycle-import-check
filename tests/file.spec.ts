import { listAllFile, resolveFilePath } from "../src/file";
import 'jest'
import { fileList, testfile2AbsPath, testfile5AbsPath } from "./test.base";

describe('file list & resolve tests', () => {

  test('should list all files', () => {
    const files = listAllFile("tests/testproject", ["js", "ts"])
    expect(files).toEqual(fileList)
  })

  test('should resolve path', () => {
    const path = resolveFilePath(testfile5AbsPath, "d2/testfile2.js")
    expect(path).toEqual(testfile2AbsPath)
  })

})
