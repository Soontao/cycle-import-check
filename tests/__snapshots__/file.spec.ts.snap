// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file list & resolve tests should correctly resolve same name 1`] = `
{
  "links": [],
  "nodes": [
    {
      "name": "tests/testproject8/express.js",
    },
    {
      "name": "tests/testproject8/m2.ts",
    },
  ],
}
`;

exports[`file list & resolve tests should map result to report VO 1`] = `
{
  "links": [
    {
      "source": "tests/testproject2/func.js",
      "target": "tests/testproject2/module1/index.js",
      "value": "import { some } from "./module1"",
    },
    {
      "source": "tests/testproject2/func.js",
      "target": "tests/testproject2/module1/part1.js",
      "value": "import { test } from "./module1/part1";",
    },
    {
      "source": "tests/testproject2/module1/index.js",
      "target": "tests/testproject2/module1/part1.js",
      "value": "export * from "./part1"",
    },
    {
      "source": "tests/testproject2/module1/index.js",
      "target": "tests/testproject2/module1/part2.js",
      "value": "export * from "./part2"",
    },
  ],
  "nodes": [
    {
      "name": "tests/testproject2/func.js",
    },
    {
      "name": "tests/testproject2/module1/index.js",
    },
    {
      "name": "tests/testproject2/module1/part1.js",
    },
    {
      "name": "tests/testproject2/module1/part2.js",
    },
  ],
}
`;
